contracts/token/ERC20/extensions/ERC3643/ERC3643.sol

```diff
@@ -0,0 +1,327 @@
+        for (uint256 i; i < fromList.length; ++i) {
+            burn(fromList[i], values[i]);
+        }
+    }
+
+    /// @inheritdoc IERC3643
+    function batchSetAddressFrozen(address[] calldata accounts, bool[] calldata frozen) public virtual {
+        for (uint256 i; i < accounts.length; ++i) {
+            setAddressFrozen(accounts[i], frozen[i]);
+        }
+    }
+
+    /// @inheritdoc IERC3643
+    function batchFreezePartialTokens(address[] calldata accounts, uint256[] calldata values) public virtual {
+        for (uint256 i; i < accounts.length; ++i) {
+            freezePartialTokens(accounts[i], values[i]);
+        }
+    }
+
+    /// @inheritdoc IERC3643
+    function batchUnfreezePartialTokens(address[] calldata accounts, uint256[] calldata values) public virtual {
+        for (uint256 i; i < accounts.length; ++i) {
+            unfreezePartialTokens(accounts[i], values[i]);
+        }
+    }
+
+    /// @inheritdoc ERC20
+    function _setName(string memory name_) internal virtual override {
+        super._setName(name_);
+        _emitUpdatedTokenInformation();
+    }
+
+    /// @inheritdoc ERC20
+    function _setSymbol(string memory symbol_) internal virtual override {
+        super._setSymbol(symbol_);
+        _emitUpdatedTokenInformation();
+    }
+
+    /**
+     * @dev Internal version of {IERC3643-setOnchainID}.
+     */
+    function _setOnchainID(address onchainID_) internal virtual {
+        _onchainID = onchainID_;
+        _emitUpdatedTokenInformation();
+    }
+
+    function _setAddressFrozen(address account, bool frozen, address owner_) internal virtual {
+        _frozen[account] = frozen;
+        emit AddressFrozen(account, frozen, owner_);
+    }
+
+    function _freezePartialTokens(address account, uint256 value) internal virtual {
+        // TODO: Unchecked block?
+        uint256 toFreeze = Math.min(value, balanceOf(account) - getFrozenTokens(account));
+        _frozenTokens[account] += toFreeze;
+        emit TokensFrozen(account, toFreeze);
+    }
+
+    function _unfreezePartialTokens(address account, uint256 value) internal virtual {
+        // TODO: Unchecked block?
+        _frozenTokens[account] -= value;
+        emit TokensUnfrozen(account, value);
+    }
+
+    function _setIdentityRegistry(IIdentityRegistry identityRegistry_) internal virtual {
+        _identityRegistry = identityRegistry_;
+        emit IdentityRegistryAdded(identityRegistry_);
+    }
+
+    function _setCompliance(ICompliance compliance_) internal virtual {
+        _compliance = compliance_;
+        emit ComplianceAdded(compliance_);
+    }
+
+    function _forcedTransfer(address from, address to, uint256 value) internal virtual returns (bool) {
+        if (identityRegistry().isVerified(to)) {
+            // TODO: Unchecked block?
+            uint256 freeBalance = balanceOf(from) - getFrozenTokens(from);
+            if (value > freeBalance) {
+                _unfreezePartialTokens(from, value - freeBalance);
```

Comment:
Wouldn't it be cheaper to:
1. update
2. if frozen > balance then set frozen to balance
?

Reply:
Yeah that makes sense.